spec/java/util/stream/Stream.main.lsl

```diff
@@ -0,0 +1,1049 @@
+            _allMatch_loop(i, predicate, result)
+        );
+    }
+
+
+    @Phantom proc _allMatch_loop (i: int, predicate: Predicate, result: boolean): void
+    {
+        if (!action CALL(predicate, [this.storage[i]]))
+        {
+            result = false;
+            action LOOP_BREAK();
+        }
+    }
+
+
+    fun *.noneMatch (@target self: Stream, predicate: Predicate): boolean
+    {
+        if (predicate == null)
+            _throwNPE();
+
+        result = true;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, this.length, +1,
+            _noneMatch_loop(i, predicate, result)
+        );
+    }
+
+
+    @Phantom proc _noneMatch_loop (i: int, predicate: Predicate, result: boolean): void
+    {
+        if (action CALL(predicate, [this.storage[i]]))
+        {
+            result = false;
+            action LOOP_BREAK();
+        }
+    }
+
+
+    fun *.findFirst (@target self: Stream): Optional
+    {
+        result = _findFirst();
+    }
+
+
+    fun *.findAny (@target self: Stream): Optional
+    {
+        result = _findFirst();
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.iterator (@target self: Stream): Iterator
+    {
+        result = new StreamIteratorAutomaton(state = Initialized,
+            parent = self,
+            cursor = 0
+        );
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.spliterator (@target self: Stream): Spliterator
+    {
+        result = action DEBUG_DO("java.util.Spliterators.spliterator(this.storage, Spliterator.ORDERED)");
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.isParallel (@target self: Stream): boolean
+    {
+        result = this.isParallel;
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.sequential (@target self: Stream): BaseStream
+    {
+        this.isParallel = false;
+        result = self;
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.parallel (@target self: Stream): BaseStream
+    {
+        this.isParallel = true;
+        result = self;
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.unordered (@target self: Stream): BaseStream
+    {
+        result = self;
+    }
+
+
+    // within java.util.stream.BaseStream
+    fun *.onClose (@target self: Stream, arg0: Runnable): BaseStream
+    {
+        val listLength: int = action LIST_SIZE(this.closeHandlers);
+        action LIST_INSERT_AT(this.closeHandlers, listLength, arg0);
+        result = self;
+    }
+
+
+    // within java.lang.AutoCloseable
+    fun *.close (@target self: Stream): void
+    {
+        val listLength: int = action LIST_SIZE(this.closeHandlers);
+
+        // NOTE: this implementation does not care about suppressing and throwing exceptions produced by handlers
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, listLength, +1,
+            _closeHandlers_loop(i)
+        );
+
+        this.closeHandlers = action LIST_NEW();
```

Comment:
Is this possible to use (call other methods, like `min/max`) this stream after closing it?

Reply:
@dog-m No - it will be thrown exception "stream has already been operated upon or closed";
What I can do with this ? Add boolean flag "closed" to StreamAutomaton ?

Reply:
I agree with using `closed` flag idea.